kerchunk/hdf.py

```diff
@@ -89,8 +94,17 @@ def __init__(
             fs, path = fsspec.core.url_to_fs(h5f, **(storage_options or {}))
             self.input_file = fs.open(path, "rb")
             url = h5f
-        else:
+            self._h5f = h5py.File(self.input_file, mode="r")
+        elif isinstance(h5f, io.IOBase):
             self.input_file = h5f
+            self._h5f = h5py.File(self.input_file, mode="r")
+        elif isinstance(h5f, (h5py.File, h5py.Group)):
+            # assume h5py object (File or group/dataset)
+            self._h5f = h5f
+            fs, path = fsspec.core.url_to_fs(url, **(storage_options or {}))
+            self.input_file = fs.open(path, "rb")
```

Comment:
_h5f is indeed set to the input two lines above. This exists for any inlining that might happen, which requires getting bytes directly from the original file, not going via h5py. 

> mess up my use case

What happens? I think providing the URL/options will certainly be required.

Reply:
in my case it's looking for a local file even if I pass valid S3 `storage_options` - leave it like this for now, I'll need to do a wee bit more testing to understand what's going on, and will get back to you if Kerchunk needs changing :+1:

Reply:
The urls starts with "s3://"?

Reply:
yes and no :rofl: It's a very peculariar bucket, the storage options dict that s3fs recognizes is
```
{'key': 'xxxx', 'secret': "xxxx", 'client_kwargs': {'endpoint_url': 'https://uor-aces-o.s3-ext.jc.rl.ac.uk'}, 'default_fill_cache': False, 'default_cache_type': 'first'}
```
the call to `s3fs` to able to read such a strange bucket is as follows:
```python
fs = s3fs.S3FileSystem(**storage_options)
with fs.open(file_url, 'rb') as s3file:
...
```

**but** `file_url` needs to be the truncated (bucket + file-name) ie `bnl/da193a_25_day__198807-198807.nc` in this case, and s3fs is assembling its full URL via the endpoint URL and that truncated bucket _ filename - it's odd, not 100% sure why this type of s3 storage wants that configuration, but bottom line is in the case of Kerchunk trying to open it as a regular s3 file it's not working - even if I prepend a correct full s3://...path to the file, I get Forbidden access since the storage identification is done wrongly

Reply:
disclaimer: I know close to diddlysquat about S3 storages, if ain't POSIX, am not flying, but heyho - I gotta support that type of storage too

Reply:
so your `fsspec.core.url_to_fs` will assemble a local fs by looking at a file url like `bnl/da193a_25_day__198807-198807.nc` but if I don't pass that, and instead I give you the full thing `s3://uor-aces-o.s3-ext.jc.rl.ac.uk/bnl/da193a_25_day__198807-198807.nc`, with the `storage_options` dict I have, the fs will indeed be assembled via s3fs, but the S3 storage I got to talk to will toss a Forbidden error since it'll look for a file named `s3://uor-aces-o.s3-ext.jc.rl.ac.uk/bnl/da193a_25_day__198807-198807.nc` and not the existing `bnl/da193a_25_day__198807-198807.nc` - probably because this started initially as an anon=True bucket, onto which we added a key and secret, just to be able to test access via valid credentials, or maybe just this S3 store is wonky, dunno :grin:

Reply:
> s3://uor-aces-o.s3-ext.jc.rl.ac.uk/bnl/da193a_25_day__198807-198807.nc

This is definitely not the right URL: the first part should be the bucket, not a server name (I'm surprised it even attempts to connect). The URL should be "s3://bnl/da193a_25_day__198807-198807.nc", as the server/endpoint is already included in the storage options.

Reply:
blast! That worked! I knew I'm not doing something right :laughing:

Reply:
though am getting fairly long times from `visititems()` - very much comparable times to the ones where there is no kerchunking done on a single Group, but rather, on the entire file

Reply:
ah that's because this `self._h5f = h5py.File(self.input_file, mode="r")` is a few lines down :grin: